File: 01576. Replace All 's to Avoid Consecutive Repeating Characters/solution.ts
/**
 * Problem: 1576. Replace All ?'s to Avoid Consecutive Repeating Characters
 *
 * Difficulty: Easy
 *
 * Language: TypeScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Replaces each '?' to avoid consecutive repeating characters
 *
 * @param s - Input string containing '?' characters
 *
 * @returns Modified string with no consecutive repeating characters
 */
const modifyString = (s: string): string => {
  // Convert string to mutable array for in-place replacement
  const chars: string[] = s.split('')

  // Iterate through each position
  for (let i = 0; i < chars.length; i++) {
    // Process only '?' placeholders
    if (chars[i] !== '?') continue

    // Try candidate letters a, b, c to find valid replacement
    for (let code = 97; code <= 99; code++) {
      // Candidate character from a to c
      const candidate: string = String.fromCharCode(code)

      // Check left neighbor is different (already replaced if it was '?')
      const leftOk: boolean = i === 0 || chars[i - 1] !== candidate

      // Check right neighbor is different (skip if right is '?' placeholder)
      const rightOk: boolean =
        i === chars.length - 1 || chars[i + 1] !== candidate

      // Pick first candidate satisfying both neighbors
      if (leftOk && rightOk) {
        // Assign valid replacement
        chars[i] = candidate
        break
      }
    }
  }

  // Join array back to string
  return chars.join('')
}

File: 01576. Replace All 's to Avoid Consecutive Repeating Characters/solution.cpp
/**
 * Problem: 1576. Replace All ?'s to Avoid Consecutive Repeating Characters
 *
 * Difficulty: Easy
 *
 * Language: C++
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

class Solution {
public:
  string modifyString(string s) {
    // Iterate through each position
    for (int i = 0; i < (int)s.size(); i++) {
      // Process only '?' placeholders
      if (s[i] != '?')
        continue;

      // Try candidate letters a, b, c to find valid replacement
      for (char c = 'a'; c <= 'c'; c++) {
        // Check left neighbor is different (already replaced if it was '?')
        bool leftOk = i == 0 || s[i - 1] != c;

        // Check right neighbor is different (skip if right is '?' placeholder)
        bool rightOk = i == (int)s.size() - 1 || s[i + 1] != c;

        // Pick first candidate satisfying both neighbors
        if (leftOk && rightOk) {
          // Assign valid replacement
          s[i] = c;
          break;
        }
      }
    }

    // Return modified string
    return s;
  }
};
